File: src/tools/sendKey.ts
import { z } from "zod";
import { TerminalManager } from "../terminal/index.js";
import { getKeySequence, getAvailableKeys } from "../utils/keys.js";

export const sendKeySchema = z.object({
  key: z.string().describe("The key to send (e.g., 'Enter', 'Tab', 'Ctrl+C', 'ArrowUp')"),
  sessionId: z.string().optional().describe("Target session ID. Omit to target the default session."),
});

export type SendKeyArgs = z.infer<typeof sendKeySchema>;

const availableKeys = getAvailableKeys();

export const sendKeyTool = {
  name: "sendKey",
  description: "Send a special key or key combination to the terminal. Common keys: Enter, Tab, Escape, Backspace, Delete, ArrowUp/Down/Left/Right, Home, End, PageUp, PageDown. Control sequences: Ctrl+C (interrupt), Ctrl+D (EOF), Ctrl+Z (suspend), Ctrl+L (clear screen), Ctrl+A (line start), Ctrl+E (line end), Ctrl+U (clear line). Function keys: F1-F12. Pass sessionId to target a specific session.",
  inputSchema: {
    type: "object" as const,
    properties: {
      key: {
        type: "string",
        description: "The key to send (e.g., 'Enter', 'Tab', 'Ctrl+C', 'ArrowUp', 'Escape')",
      },
      sessionId: {
        type: "string",
        description: "Target session ID. Omit to target the default session.",
      },
    },
    required: ["key"],
  },
};

export function handleSendKey(manager: TerminalManager, args: unknown): { content: Array<{ type: "text"; text: string }> } {
  const parsed = sendKeySchema.parse(args);
  const sequence = getKeySequence(parsed.key);

  if (sequence === null) {
    const available = getAvailableKeys();
    throw new Error(
      `Unknown key: "${parsed.key}". Available keys include: ${available.slice(0, 15).join(", ")}...`
    );
  }

  manager.write(sequence, parsed.sessionId);

  return {
    content: [
      {
        type: "text",
        text: `Sent key: ${parsed.key}`,
      },
    ],
  };
}

File: src/tools/send_key.rs
use serde::Deserialize;
use serde_json::Value;

use crate::mcp::{CallToolResult, ToolDefinition};
use crate::terminal::Terminal;

/// Arguments for the sendKey tool
#[derive(Debug, Deserialize)]
pub struct SendKeyArgs {
    /// The key to send (e.g., "Enter", "Tab", "Ctrl+C", "Up")
    pub key: String,
}

/// Send a special key or key combination to the terminal
pub fn send_key(terminal: &mut Terminal, args: Value) -> Result<CallToolResult, String> {
    let args: SendKeyArgs = serde_json::from_value(args)
        .map_err(|e| format!("Invalid arguments: {}", e))?;

    terminal
        .send_key(&args.key)
        .map_err(|e| format!("Failed to send key: {}", e))?;

    Ok(CallToolResult::text(format!("Sent key: {}", args.key)))
}

/// Get the tool definition for sendKey
pub fn definition() -> ToolDefinition {
    ToolDefinition {
        name: "sendKey".to_string(),
        description: "Send a special key or key combination to the terminal. Supports keys like Enter, Tab, Escape, arrow keys (Up, Down, Left, Right), function keys (F1-F12), and control combinations (Ctrl+C, Ctrl+D, etc.).".to_string(),
        input_schema: serde_json::json!({
            "type": "object",
            "properties": {
                "key": {
                    "type": "string",
                    "description": "The key to send. Examples: 'Enter', 'Tab', 'Escape', 'Up', 'Down', 'Left', 'Right', 'Ctrl+C', 'Ctrl+D', 'Ctrl+Z', 'F1', 'Home', 'End', 'PageUp', 'PageDown', 'Backspace', 'Delete'"
                }
            },
            "required": ["key"]
        }),
    }
}
